### DISCUSSION TITLE
単体テストの考え方/使い方 7章

### DISCUSSION THREAD

**Discussion by y-kaneda** [atman-inc/training#102] [General]
輪読会のサマリ用ディスカッション。
7章分の各章単位程度でコメントを記載するので、不明点、相談点、確認点、勉強になったポイントなどは、各自ディスカッションのスレッドへコメントをお願いします。

**Comment by y-kaneda** [comment]
# 第7章 単体テストの価値を高めるリファクタリング

**Reply by y-kaneda** [reply]
- 価値のあるテスト・ケースを作成できるようにする
- プロダクション・コードの質が高くなければ価値のあるテスト・ケースは作成できない
  - よくある「テストしづらいコード」

**Comment by y-kaneda** [comment]
## 7.1 リファクタリングが必要なコードの識別

**Reply by nmurakami0** [reply]
テストコードとプロダクションコードは本質的に影響しあうので、テストの改善にはプロダクションコードに大抵手をつける必要がある。

**Comment by y-kaneda** [comment]
### 7.1.1 4種類のプロダクション・コード

**Reply by y-kaneda** [reply]
- コードの複雑さ
  - 分岐の数
  - 循環的複雑度(cylomatic complexity)
- ドメインにおける重要性
  - ドメイン層のコードはエンドユーザの求めていることに直接結びつく
- 複雑 ≠ 重要
  - 重要・緊急のマトリクス的な勘違いはありそう
- 協力者オブジェクトの数
  - 協力者オブジェクトが多いとテスト・ケースのコード量は増え、保守のしやすさは下がる
  - モックも増える
- 複雑さ・重要度と協力者オブジェクトの2軸による4象限で分類する
- テスト網羅率100%を目指すのではなく、価値のあるテストを目指す

**Reply by xxManatoxx** [reply]
- 分類する視点
  - コードの複雑さ、もしくはドメインにおける重要性
  - 協力者オブジェクトの数
- 2つの分類から見た時に4種類に分類したプロダクションコード
  - ドメインモデル/アルゴリズム
  - 取るに足らないコード
  - コントローラ
  - 過度に複雑なコード

- 感想
  - テストの実装もそうですがプロダクションコードの作成時に4種類に分類されたコードのうち過度に複雑なコードにならないように気をつけたい
  - 作成段階でできる限りコントローラとドメインモデル/アルゴリズムに分けて実装できるのが理想かなと思いました
  - 上記気をつけられるようにデザインパターンの理論を理解しておくのが重要なのだと読んでいて感じました

**Reply by nmurakami0** [reply]
cyclomatic complexityの計算で注意すべきこと:
`IF 条件1 AND 条件2 THEN`は、`IF 条件1 THEN IF 条件2 THEN`と同じなので、1 + 2 = 3になる。

協力者オブジェクトの数には明示されていないものも含むので注意

図7.1について:

- 協力者オブジェクトの数を常に少なくできるわけではない
  - 大して重要じゃないけれど、コントローラーは協力者オブジェクトの数が必然的に多くなる
  - ドメイン・モデル/アルゴリズムは、重要だけど、だからといって協力者オブジェクトの数が多いのは理想的ではない
- 右上の状態になっていたら、アルゴリズムとコントローラーに分割すると良い

感想: 

1つのクラス・メソッドで色々やりすぎているので引数で結果を渡すようにする、というのは良く行うと思うが、
それもこの分割に該当しそう。

**Comment by y-kaneda** [comment]
### 7.1.2 質素なオブジェクト(Humble Object)を用いた過度に複雑なコードの分割

**Reply by xxManatoxx** [reply]
- 質素なオブジェクト
  - 関数型アーキテクチャ、ヘキサゴナルアーキテクチャで実際に適用していた
  - 各クラスが持てる責任を一つだけにするという単一責任の原則を遵守するための手段
  - 単一責任に該当するものがビジネスロジック

**Reply by nmurakami0** [reply]
Humble Objectパターン

- テストを行うことが難しい依存をもたせ、ロジックは持たせない
- PresenterやControllerはHumble Objectの一例
- ヘキサゴナルアーキテクチャと関数型アーキテクチャはこのパターンを適用したもの

**Comment by y-kaneda** [comment]
## 7.2 単体テストに価値を持たせるためのリファクタリング

**Comment by y-kaneda** [comment]
### 7.2.1 サンプル・プロジェクト(ユーザ管理システム)について

**Reply by y-kaneda** [reply]
- なんで public 変数に代入するの
- DBから取ってきた値、Userオブジェクトとかじゃなくてただの配列なの気になる
  - でもORマッパー使わないとこうなるか
- int delta ってどういう命名
- ドメインロジックに Database と MessageBus という暗黙的な依存(プロセス外依存)が入っている
- Active Record パターン
  - ドメイン・クラスがデータベースを扱う
  - シンプルで寿命が短いプロジェクトだと良いが、責務の分離ができず、肥大化していくと厳しくなる

**Reply by nmurakami0** [reply]
1メソッドに書くと確かにこうなりそう。Userクラスには書かないと思いますが。
キャストが多いのが地味に気になる。

**Comment by y-kaneda** [comment]
### 7.2.2 今回は行わないリファクタリング：暗黙的な依存を明示された依存にする

**Reply by y-kaneda** [reply]
- DatabaseクラスとMessageBusクラスをDIするとモックできるようになるが、ドメイン・モデルがプロセス外依存を扱うことには変わりない
- テストのモック部分の保守の分、コストは増えテストは壊れやすくなる
- ヘキサゴナル・アーキテクチャではドメイン・モデルに外部システムとのコミュニケーションの責務を負わせない

**Comment by y-kaneda** [comment]
### 7.2.3 1回目のリファクタリング：アプリケーション・サービス層の導入

**Reply by nmurakami0** [reply]
まだ途中だと注釈があるが、それにしてもUserクラスのChangeEmailがnumberOfEmployeesをreturnしているのは違和感あり

**Comment by y-kaneda** [comment]
### 7.2.4 2回目のリファクタリング：アプリケーション・サービスから複雑さを取り除く

**Reply by nmurakami0** [reply]
Columnの、変換ロジックもテストすべきというのは、現実だと端折ってしまっている時もありそうと自戒。
確かにこの例だとキャストしたりしているので、例外発生など考慮すべきパターンは多い。

**Reply by y-kaneda** [reply]
- フレームワークやライブラリの内部実装は基本的には見ないが、確かにテストすべき複雑性はある

**Comment by y-kaneda** [comment]
### 7.2.5 3回目のリファクタリング：新たなCompanyクラスの導入

**Comment by y-kaneda** [comment]
## 7.3 プロダクション・コードの種類に基づく効果的な単体テストの作成

**Comment by y-kaneda** [comment]
### 7.3.1 ドメイン層とユーティリティ・コードに対するテスト

**Comment by y-kaneda** [comment]
### 7.3.2 その他の種類に分類されるコードに対するテスト

**Comment by y-kaneda** [comment]
### 7.3.3 事前条件をテストすべきか？

**Comment by y-kaneda** [comment]
## 7.4 コントローラにおける条件付きロジックの扱い

**Reply by y-kaneda** [reply]
- クリーン・アーキテクチャ的な分離のお話
- 処理の途中結果でプロセス外依存から新たにデータを取得するような場合、綺麗に分けられないことがある
- 対策3選
  - 最初に全てのデータを取得しておく
    - テストのしやすさ○、コントローラの簡潔さ○、パフォーマンス×
  - ドメイン・モデルへのプロセス外依存のDI
    - テストのしやすさ×、コントローラの簡潔さ○、パフォーマンス○
  - 決定を下す過程を細かく分ける
    - テストのしやすさ○、コントローラの簡潔さ×、パフォーマンス○
    - 本書では一番オススメ

**Reply by koushisa** [reply]
カプセル化したい重要なドメイン知識はドメインモデルに凝集させるという話

フロントエンドのケースだとSkeleton UIを実装するときが似たケースかもと思った
Skeletonの領域を親で確保するか、子で確保するか。

本書の考え方だと、Skeletonの詳細と統一感子が確定させるので、子に凝集させることを目指すと思われる。

Vercelでの実装例
  - 定義: https://github.com/vercel/app-playground/blob/14d123160619e45bacf17a2afd59fe4eb8d1ca17/app/streaming/_components/reviews.tsx#L32
  - 配置: https://github.com/vercel/app-playground/blob/14d123160619e45bacf17a2afd59fe4eb8d1ca17/app/streaming/node/product/%5Bid%5D/page.tsx#L48-L63

**Comment by y-kaneda** [comment]
### 7.4.1 確認後実行（CanExecute/Execute）パターンの適用

**Reply by y-kaneda** [reply]
- early return すればいいじゃないというようなお話

**Reply by nmurakami0** [reply]
CanChangeEmailというドメイン層のメソッドに切り出して、

- Controllerでの事前チェックに用いる
- Preconditionとして、対象の操作の際の再チェックにも用いる

ようにすればよい。
そうすることで、

- Controllerに、メールアドレスを変更出来るかというドメインロジックを配置する必要がなくなる
- 確認をせずにメールアドレスを変更することが出来ない

となる。

**Reply by koushisa** [reply]
- コントローラが何かのオブジェクトのpropertyに直でアクセスしていたら危険信号
- コントローラはオブジェクトの詳細は意識せずに、目的に必要な最小限の処理をドメインモデルへ命令すれば良い
- 目的を達成するための知識やロジックの詳細はドメインモデルに集める
- ドメインモデルの中の処理が純粋関数になっていると尚良い

**Comment by y-kaneda** [comment]
### 7.4.2 ドメイン・モデルの状態を追跡するドメイン・イベントの利用

**Reply by nmurakami0** [reply]
以前話題に出ていた、メールアドレス変更しなくてもメッセージが送信される、というバグは意図的なものだった。
その解決策としてドメインイベントを用いる。

**Reply by y-kaneda** [reply]
ドメイン・イベント
- 物理的なイベントではなく、内部的な状態の変化によって発生する
- 外部に伝えるデータを含んだクラス
- なんとなくstateとその検知っぽい

**Comment by y-kaneda** [comment]
## 7.5 結論

**Comment by y-kaneda** [comment]
## まとめ
